### PR TITLE
added binary,ternary,and quantized version of the 3-layer model

### PR DESCRIPTION
**Pull request by jngadiub**
Inspired from the MNIST-mpl example here:

https://github.com/DingKe/nn_playground/blob/master/binarynet/mnist_mlp.py

that uses fully connected layers + binary (or ternary) activation function + batch normalization.

Here the references:

binary networks: https://arxiv.org/abs/1602.02830
ternary networks: https://arxiv.org/abs/1605.04711
quantized networks: https://arxiv.org/abs/1609.07061

For quantized networks, one in principle can set whatever precision for both weights and activations. I gave a try with 4 bits as set in the PR but I can't reproduce what I get with binary or ternary. This needs to be understood.

In general, I apply a MinMaxScaler to the inputs first and re-categorize the classes such that

0 --> -1
1 --> 1

I have also changed for these cases the loss function from categorical_crossentropy to squared_hinge as used in the MNIST example. 

There is a loss of performance on the AUC of about 10%:

training_bnn

training_tnn

Need to understand also why the final weights are not binarized (or ternarized). And more tests are needed to understand the over training visible in the validation loss function versus epoch and/or if a bigger network would actually gain more from the binarization.

The batch normalization layer is a nice addition to the hls4ml library but I would not expect it to be fundamental for a binary/ternary network case where activation functions and weights are already constrained, although it is used in the references.

### CODE REVIEW

File: train/quantized_layers.py [hunk 1]
```diff
@@ -0,0 +1,580 @@
+# -*- coding: utf-8 -*-
```
Comment (jmduarte) [inline train/quantized_layers.py]: i would move this to it's own `layers/` subfolder

**Comment by jmduarte** [comment]
@jngadiub one other comment: do we want to make the default config for these networks with 10x more nodes (so that the performance is ~the same as what we have now)?

**Comment by jngadiub** [comment]
Yes, I agree with you. Let me do that.
